core/kernel/interrupt.c

```diff
@@ -35,17 +35,25 @@ static struct itr_handler *find_handler(size_t it)
 
 void itr_handle(size_t it)
 {
-	struct itr_handler *h = find_handler(it);
+	struct itr_handler *h = NULL;
+	enum itr_return ret = ITRR_NONE;
 
-	if (!h) {
-		EMSG("Disabling unhandled interrupt %zu", it);
-		itr_chip->ops->disable(itr_chip, it);
-		return;
-	}
+	SLIST_FOREACH(h, &handlers, link) {
+
+		if (h->it != it)
+			continue;
+
+		ret = h->handler(h);
 
-	if (h->handler(h) != ITRR_HANDLED) {
-		EMSG("Disabling interrupt %zu not handled by handler", it);
-		itr_chip->ops->disable(itr_chip, it);
+		if (ret == ITRR_NONE) {
```

Comment:
This assumes that the interrupt handlers knows in which order they are registered.

I'd prefer a adding another function, for instance `itr_add_shared()`, which registers its own handler which then call all the shared handlers of this interrupt.

Reply:
> >  I think you will need extra return values than ITRR_HANDLED and ITRR_NONE to handle those cases.
> 
> With our handler implementation, not belong to it's own interrupt would be return ITRR_NONE. So I think I will leave a space here for other platform for next.

Sequence should be something like: clear interrupt, call each consumer handler, check at least 1 returned HANDLED unless what interrupt can be disabled. No?

Reply:
Yes, that should work. That's a better solution than I suggested.

Reply:
> 
> 
> > > I think you will need extra return values than ITRR_HANDLED and ITRR_NONE to handle those cases.
> > 
> > 
> > With our handler implementation, not belong to it's own interrupt would be return ITRR_NONE. So I think I will leave a space here for other platform for next.
> 
> Sequence should be something like: clear interrupt, call each consumer handler, check at least 1 returned HANDLED unless what interrupt can be disabled. No?

OK , I got your point.
If all consumer not return ITRR_HANDLED , then no one can disable the interrupt, that makes the noise. 
BTW, if change the sequence to call each consumer handler , this would cost the time, because all handlers would be visited
`SLIST_FOREACH(h, &handlers, link)  `
but the original 
`find_handler() `
 just match the first hooked handler. 

so, I think @jenswi-linaro 's suggestion , 

> I'd prefer a adding another function, for instance itr_add_shared(), which registers its own handler which then call all the shared handlers of this interrupt.

is the good solution. I will modify to this and apply your opinion "to call each consumer handler" 

Many thanks,
Davidwang

Reply:
@wawang621, I preferred Etienne's solution as we worst case anyway may scan all handlers, but we're also only expecting very few handlers.

If we end up with so many handlers that performance start to suffer we should replace this singly linked list with a more advanced data structure.

Reply:
Hi @jenswi-linaro , 

I push another patch to modify as yours mention, but I am not verify it with my platform, 
Can you help to review it ? 

[https://github.com/wawang621/optee_os/commit/79ba9ffb26f9d25ff30b1d7d96e55194c17aa648](url)

Thanks,
Davidwang